libsolidity/ast/ASTUtils.h

```diff
@@ -48,4 +51,6 @@ Type const* type(Expression const& _expression);
 /// (this can happen for variables with non-explicit types before their types are resolved)
 Type const* type(VariableDeclaration const& _variable);
 
+u256 storageBaseLocationValue(ContractDefinition const& _contract);
```

Comment:
This needs a docstring. There are quite a few corner cases where it's not obvious what the function will return and the docstring should clarify that:
- The same contract can have many different layouts depending on what inheritance hierarchy it's a part of. Does the function assume that it's a top-level contract?
- Contracts without layout specifier.
- Contracts that inherit from other contracts that already have storage variables. Do we point at contract's own state variables or are inherited contracts included too?
- Contracts that can have storage but are empty.
- Contracts that can't have storage at all (i.e. libraries/interfaces).

If the function does not support any of these cases it should also have asserts against them.

Reply:
Also, maybe the names should be adjusted so it's easier to realize some of the assumptions we're making here:
```suggestion
u256 storageLayoutBaseForInheritanceHierarchy(ContractDefinition const& _topLevelContract);
```

Reply:
It should only produce a value different from the default zero if the contract has an storage base specified. In any other case, including invalid ones like interfaces and libraries, it returns zero.
At the time I was just thinking of using it in places after type checking already happened and the value exists and is set or then the contract does not specifies a base value and/or it is not the most derived one.
Not sure if it is just better to have the `if` directly in `ContractType::stateVariables` where it is used...

Reply:
It's a good idea to encapsulate it there. It should just be properly documented and have a clear name.